internal/buildengine/updates.go

```diff
@@ -0,0 +1,110 @@
+package buildengine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"connectrpc.com/connect"
+
+	buildenginepb "github.com/block/ftl/backend/protos/xyz/block/ftl/buildengine/v1"
+	enginepbconnect "github.com/block/ftl/backend/protos/xyz/block/ftl/buildengine/v1/buildenginepbconnect"
+	langpb "github.com/block/ftl/backend/protos/xyz/block/ftl/language/v1"
+	ftlv1 "github.com/block/ftl/backend/protos/xyz/block/ftl/v1"
+	"github.com/block/ftl/common/builderrors"
+	ftlerrors "github.com/block/ftl/common/errors"
+	"github.com/block/ftl/internal/channels"
+	"github.com/block/ftl/internal/log"
+	"github.com/block/ftl/internal/rpc"
+)
+
+type updatesService struct {
+	engine *Engine
+}
+
+var _ enginepbconnect.BuildEngineServiceHandler = &updatesService{}
+
+func (e *Engine) startUpdatesService(ctx context.Context, endpoint *url.URL) error {
+	logger := log.FromContext(ctx).Scope("build:updates")
+
+	svc := &updatesService{
+		engine: e,
+	}
+
+	logger.Debugf("Build updates service listening on: %s", endpoint)
+	err := rpc.Serve(ctx, endpoint,
+		rpc.GRPC(enginepbconnect.NewBuildEngineServiceHandler, svc),
+	)
+
+	if err != nil {
+		return fmt.Errorf("build updates service stopped serving: %w", err)
+	}
+	return nil
+}
+
+func (u *updatesService) Ping(context.Context, *connect.Request[ftlv1.PingRequest]) (*connect.Response[ftlv1.PingResponse], error) {
+	return connect.NewResponse(&ftlv1.PingResponse{}), nil
+}
+
+func (u *updatesService) StreamEngineEvents(ctx context.Context, req *connect.Request[buildenginepb.StreamEngineEventsRequest], stream *connect.ServerStream[buildenginepb.StreamEngineEventsResponse]) error {
+	events := make(chan *buildenginepb.EngineEvent, 64)
```

Comment:
I wonder about this case:
- Start `ftl dev` in a terminal
- Build engine fails to build a module, has code errors
- IDE launches, connects to build engine for updates
- IDE thinks everything is all good, rather than receiving the initial state

Reply:
Yeah this will def happen. And actually did happen before as well. :)

We might want to think about the lsp itself being about to trigger a build when it activates to get the current errors. I'm guessing other extensions might do something similar to get their initial state 🤔